device/src/ota.rs

```diff
@@ -0,0 +1,521 @@
+        let mut read = [0u8; ESP32_OTADATA_SIZE as usize + FS_PARTITION_METADATA_SIZE as usize];
+        flash
+            .read(self.otadata_offset + (slot as u32) * 4096, &mut read)
+            .unwrap();
+
+        assert_eq!(read, bytes, "otadata read should be what was written");
+    }
+
+    pub fn start_upgrade(
+        &self,
+        flash: &mut FlashStorage,
+        size: u32,
+        expected_digest: FirmwareDigest,
+        active_partition_digest: FirmwareDigest,
+    ) -> FirmwareUpgradeMode {
+        let slot = self.next_slot(flash);
+        let partition = self.ota_partitions[slot as usize];
+        assert_eq!(
+            partition.size, FIRMWARE_IMAGE_SIZE,
+            "partition size and FIRMWARE_PAD_LENGTH must be the same"
+        );
+        assert!(
+            partition.size % FIRMWARE_UPGRADE_CHUNK_LEN == 0,
+            "these should match up to avoid overwriting"
+        );
+
+        if expected_digest == active_partition_digest {
+            FirmwareUpgradeMode::Passive {
+                size,
+                sent_ack: false,
+            }
+        } else {
+            FirmwareUpgradeMode::Upgrading {
+                ota: *self,
+                ota_slot: slot,
+                expected_digest,
+                size,
+                state: State::WaitingForConfirm { sent_prompt: false },
+            }
+        }
+    }
+}
+
+pub enum FirmwareUpgradeMode {
+    Upgrading {
+        ota: OtaConfig,
+        ota_slot: u8,
+        expected_digest: FirmwareDigest,
+        size: u32,
+        state: State,
+    },
+    Passive {
+        size: u32,
+        sent_ack: bool,
+    },
+}
+
+#[derive(Clone, Debug)]
+pub enum State {
+    WaitingForConfirm { sent_prompt: bool },
+    Erase { seq: u32 },
+    WaitingToEnterUpgradeMode,
+}
+
+impl FirmwareUpgradeMode {
+    pub fn poll(
+        &mut self,
+        flash: &mut FlashStorage,
+    ) -> (Option<DeviceSendBody>, Option<ui::Workflow>) {
+        match self {
+            FirmwareUpgradeMode::Upgrading {
+                ota,
+                ota_slot,
+                expected_digest,
+                size,
+                state,
+            } => {
+                let partition = ota.ota_partitions[*ota_slot as usize];
+                match state {
+                    State::WaitingForConfirm { sent_prompt } if !*sent_prompt => {
+                        *sent_prompt = true;
+                        (
+                            None,
+                            Some(ui::Workflow::UserPrompt(
+                                ui::Prompt::ConfirmFirmwareUpgrade {
+                                    firmware_digest: *expected_digest,
+                                    size: *size,
+                                },
+                            )),
+                        )
+                    }
+                    State::Erase { seq } => {
+                        let mut finished = false;
+                        for _ in 0..32 {
```

Comment:
Is this some maximum number of sectors per image?

Reply:
Good question. No it was just to make it go faster by updating the screen less. I'll see if it's still worthwhile.